pkgs/extension_discovery/README.md

```diff
@@ -0,0 +1,216 @@
+# Package Extension Discovery
+
+A convention to allow other packages to provide extension for your package.
+Including logic for finding extensions that honors this convention.
+
+The convention implemented in this package is that if `foo` provides an
+extension for `<targetPackage>`.
+Then `foo` must contain a config file `extension/<targetPackage>/config.json`.
+This file indicates that `foo` provides and extension for `<targetPackage>`.
+
+If `<targetPackage>` accepts extensions from other packages it must:
+ * Find extensions using `findExtensions('<targetPackage>')` from this package.
+ * Document how extensions are implemented:
+   * What should the contents of the `extension/<targetPackage>/config.json` file be?
+   * Should packages providing extension have a dependency constraint on `<targetPackage>`?
+   * What libraries/assets should packages that provide extension include?
+   * Should packages providing extensions specify a [topic in `pubspec.yaml`][1]
+     for easy discovery on pub.dev.
+
+The `findExtensions(targetPackage, packageConfig: ...)` function will:
+ * Load `.dart_tool/package_config.json` and find all packages that contain a
+   valid JSON file: `extension/<targetPackage>/config.json`.
```

Comment:
nit on naming of `targetPackage`. DevTools is not really a "package". Maybe instead of `targetPackage`, something like `extendableTarget`?

Reply:
This just aligns with the convention that guides usage of `.dart_tool/`:
https://dart.dev/tools/pub/package-layout#project-specific-caching-for-tools

Obviously, not everyone honors this, and the convention was just something I made up :see_no_evil: 

But if we encourage the use of a package name from pub.dev, then we reduce the risk of conflicts.

But maybe we should include a line like:
> If your tool isn’t distributed through the [pub.dev site,](https://pub.dev/) you might consider publishing a placeholder package in order to reserve the unique name.

Again, there is nothing stopping anyone from using any name, even a package name they don't own, or a string that's not a valid package name.